Add Wigner-Eckart theorem kernel and extract ci information from mc objects

bogdanoff opened the pull request:
The implementation more or less follows https://doi.org/10.1063/1.4927785 and ensures consistency with original implementation with respect to SO energies. 

With WE theorem it is much easier to apply SISO, as only one state is needed.
For now the external py3nj library is expected to be present.

hczhai commented:
It is not a good idea to add the dependence of ``py3nj``. In fact I cannot install ``py3nj`` in one of my computer environment. If you look at the issue pages of that repo you will see lots of installation problems.

The CG coefficients can be implemented in pure Python with only 30 lines, so I added the Python implementation and removed the dependence (see https://github.com/hczhai/fci-siso/pull/1/commits/602e68c9bacaa771872f248c16994d725e8eeea6). The speed of this Python implementation is 20 times faster than that of ``py3nj``:

```python3
>>> %%timeit
>>> su2cg.clebsch_gordan(5, 2, 7, -1, 2, 1)
>>> su2cg.clebsch_gordan(6, 2, 8, -4, 2, -2)
>>> su2cg.clebsch_gordan(3, 1, 2, -1, -1,-2)
17.3 µs ± 4.21 µs per loop (mean ± std. dev. of 7 runs, 100000 loops each)

>>> %%timeit
>>> py3nj.clebsch_gordan(5, 2, 7, -1, 2, 1)
>>> py3nj.clebsch_gordan(6, 2, 8, -4, 2, -2)
>>> py3nj.clebsch_gordan(3, 1, 2, -1, -1, -2)
381 µs ± 18.7 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

Otherwise it looks good to me. Please let me know if you are okay with these changes.

In ``tests/12-ru-casscf.py`` (and possibly some other tests) ideally we may need to adjust the weight of states using their multiplicities:

```python3
nstates = sum((sp + 1) * st for sp, st in zip(mySpins, myStates))
weights = [(sp + 1) / nstates for sp, st in zip(mySpins, myStates) for _ in range(st)]
```

bogdanoff commented:
> It is not a good idea to add the dependence of `py3nj`. In fact I cannot install `py3nj` in one of my computer environment. If you look at the issue pages of that repo you will see lots of installation problems.

Indeed, I  considered that as a temporary solution., though `pip install py3nj==0.1.2` worked for me in all cases.

> The CG coefficients can be implemented in pure Python with only 30 lines, so I added the Python implementation and removed the dependence (see [602e68c](https://github.com/hczhai/fci-siso/commit/602e68c9bacaa771872f248c16994d725e8eeea6)). The speed of this Python implementation is 20 times faster than that of `py3nj`
> Otherwise it looks good to me. Please let me know if you are okay with these changes.

Thanks a lot, that is great! Definitely I'm fine with your changes.

> In `tests/12-ru-casscf.py` (and possibly some other tests) ideally we may need to adjust the weight of states using their multiplicities:
> 
> ```python
> nstates = sum((sp + 1) * st for sp, st in zip(mySpins, myStates))
> weights = [(sp + 1) / nstates for sp, st in zip(mySpins, myStates) for _ in range(st)]
> ```

Indeed, that is the choice through the examples in this repo and I understand why it can be done. But in practice I most of the times use the spin-free "democratic" state-averaging giving equal weight to each spin-free state.
